Updated Platform and Plugin Versioning (#979)
apache/cordova-docs: www/docs/en/dev/platform_plugin_versioning_ref/index.md

The 'restore' step happens automatically when a **`cordova prepare`** is issued, making use of information previously saved in the config.xml and package.json files.

One scenario where save/restore capabilities come in handy is in large teams that work on an app, with each team member focusing on a platform or plugin. This feature makes it easier to share the project and reduce the amount of redundant code that is checked in the repository.

## Platform Versioning

### Saving Platforms

To save a platform, issue the following command:

```bash
cordova platform add <platform[@<version>] | directory | git_url>
```

After running the above command, the **`package.json`** should contain something as seen below:

```json
"cordova": {
  "platforms": [
    "android"
  ]
},

> "dependencies": {

Reviewer: Is it `dependencies` or `devDependencies`? It really *should* be `devDependencies` but I suspect not...

Author: When looking at a project that I created awhile back with Cordova 9, it put the platform as a `dependencies` and plugins as `devDependencies`.

Reviewer: Yay consistency :roll_eyes:

  "cordova-android": "^8.0.0",
}
```

> The `--nosave` flag prevents adding and deleting specified platforms to the `package.json` file. To prevent saving a platform, issue the following command:

Reviewer: FYI the `--nosave` flag is known to have issues, including the removal of unrelated packages from the node_modules folder.

**`config.xml`**:

```xml
<engine name="android" spec=“7.4.1” />
```

**`package.json`**:

```json
"cordova": {
  "platforms": [
    "android"
  ]
},
"dependencies": {
  "cordova-android": "^8.0.0"
}
```

When `prepare` is executed, the version from `package.json` has higher priority over  `config.xml` and version `^8.0.0` will be installed.

---

## Plugin Versioning

The plugin commands are a mirror of the platform commands

### Saving plugins

To save a plugin, you issue the following command:

```bash
cordova plugin add <plugin[@<version>] | directory | git_url>
```

After running the above command, the **`package.json`** should contain something as seen below:

```json
"cordova": {
  "plugins": [
    "cordova-plugin-device"
  ]
},
"devDependencies": {
  "cordova-plugin-device": "^1.0.0"
}
```

The `--nosave` flag prevents adding and deleting specified plugins from `package.json`. To prevent saving a plugin, you issue the following command:

```bash
cordova plugin add <plugin[@<version>] | directory | git_url> --nosave
```

Some Examples:

* **`cordova plugin add cordova-plugin-device`**

>   Retrieves the pinned version of the `cordova-plugin-device` plugin from NPM, adds it to the project and updates the `package.json` file.

Reviewer: npm is never capitalized

* **`cordova plugin add https://github.com/apache/cordova-plugin-device.git`**

  NPM retrieves the `cordova-plugin-device` plugin from the git repository, adds it to the project and updates the `package.json`.

* **`cordova plugin add C:/path/to/console/plugin`**

  Retrieves the `cordova-plugin-device` plugin from the specified directory, adds it to the project, and updates the `package.json` file.

### Mass Saving of Plugins on an Existing Project

If you have a pre-existing project and you want to save all currently added plugins in the project, you can use:

```bash
cordova plugin save
```

### Removing Plugins

It is also possible to delete from config.xml and package.json during the command `cordova plugin remove`:

```bash
cordova plugin remove <plugin>
```

For Example:

* **`cordova plugin remove cordova-plugin-device`**

  Removes the `cordova-plugin-device` plugin from the project and deletes its entry from `package.json`.

  _Note: If the plugin definiton existed in `config.xml` from a previous version of Cordova-CLI, it will also be removed from `config.xml`._

### Restoring Plugins

> Plugins are automatically restored from the `package.json` and `config.xml` when executing the the **`cordova prepare`** command.

Reviewer: Adding "the" seems a bit weird here unless you also add "files" after

> If a plugin is defined in both files, the information defined in `package.json` is used as the soruce of truth.

Reviewer: Typo: soruce -> source

One scenario where save/restore capabilities come in handy is in large teams that work on an app, with each team member focusing on a platform or plugin. This feature makes it easier to share the project and reduce the amount of redundant code that is checked in the repository.

## Platform Versioning

### Saving Platforms

To save a platform, issue the following command:

```bash
cordova platform add <platform[@<version>] | directory | git_url>
```

After running the above command, the **`package.json`** should contain something as seen below:

```json
"cordova": {
  "platforms": [
    "android"
  ]
},
"dependencies": {
  "cordova-android": "^8.0.0",
}
```

The `--nosave` flag prevents adding and deleting specified platforms to the `package.json` file. To prevent saving a platform, issue the following command:

```bash
cordova platform add <platform[@<version>] | directory | git_url> --nosave
```

Some Examples:

* **`cordova platform add android`** 

  Retrieves the pinned version of `cordova-android` platform from npm, adds it to the project and updates the `package.json` file.

> 

Reviewer 2: Not a thing anymore?

Author: Yeah been removed since Dec 2017 in this commit: https://github.com/apache/cordova-lib/commit/33d84dd734a47d1d2d9bd85b6e6017b3c59bfbe5

* **`cordova platform add android@7.1.4`**

  Retrieves the `cordova-android` platform version `7.1.4` from npm, adds it to the project and updates the `package.json` file.

> * **`cordova platform add https://github.com/apache/cordova-android.git`**

Reviewer 2: ```suggestion
* **`cordova platform add https://github.com/apache/cordova-android.git`**, **`cordova platform add https://github.com/apache/cordova-android`** or **`cordova platform add github:apache/cordova-android`**
```

  npm retrieves the `cordova-android` platform from the git repository, adds it to the project and updates the `package.json`.
  
* **`cordova platform add C:/path/to/android/platform`**

  Retrieves the Android platform from the specified directory, adds it to the project, and updates the `package.json` file.

* **`cordova platform add android --nosave`**

  Retrieves the pinned version of `cordova-android` platform from npm, adds it to the project, but does not add it to the `package.json` file.

### Updating or Removing Platforms

> It is also possible to update or delete from config.xml and `package.json` during the commands 'cordova platform update' and 'cordova platform remove' :

Reviewer 2: ```suggestion
It is also possible to update or delete from `config.xml` and `package.json` during the commands 'cordova platform update' and 'cordova platform remove' :
```

  Removes the `cordova-android` platform from the project, but does not remove it from the `package.json` file.

### Restoring Platforms

Platforms are automatically restored from the `package.json` and `config.xml` when executing the the **`cordova prepare`** command.

If a platfrom is defined in both files, the information defined in `package.json` is used as the source of truth.

After `prepare`, if a platform was restored from `config.xml`, the `package.json` file will updated to reflect the restored platform and will have identical vales from `config.xml`.

If you add a platform without specifying a `<version | folder | git_url>`, the version that will be installed is taken from `package.json` or `config.xml`.

**If discovered** in both files, `package.json` is given higher priority over `config.xml`.

Example:

Suppose your `config.xml` file contains the following entry:

```xml
<?xml version='1.0' encoding='utf-8'?>
    ...
    <engine name="android" spec="7.1.4" />
    ...
</xml>
```

If you run the command **`cordova platform add android`** with no `<version | folder | git_url>` specified, the platform `android@7.1.4` will be retrieved and installed.

> **Example Order of Priorty for Restoring Platforms:**

Reviewer 2: ```suggestion
**Example Order of Priority for Restoring Platforms:**
```

---

## Plugin Versioning

The plugin commands are a mirror of the platform commands

### Saving Plugins

To save a plugin, you issue the following command:

```bash
cordova plugin add <plugin[@<version>] | directory | git_url>
```

After running the above command, the **`package.json`** should contain something as seen below:

```json
"cordova": {
  "plugins": [
    "cordova-plugin-device"
  ]
},
"devDependencies": {
  "cordova-plugin-device": "^1.0.0"
}
```

The `--nosave` flag prevents adding and deleting specified plugins from `package.json`. To prevent saving a plugin, you issue the following command:

```bash
cordova plugin add <plugin[@<version>] | directory | git_url> --nosave
```

Some Examples:

* **`cordova plugin add cordova-plugin-device`**

  Retrieves the pinned version of the `cordova-plugin-device` plugin from npm, adds it to the project and updates the `package.json` file.

* **`cordova plugin add cordova-plugin-device@2.0.1`**

  Retrieves the `cordova-plugin-device` plugin at version `2.0.1` from npm, adds it to the project and updates the `package.json` file.

> * **`cordova plugin add https://github.com/apache/cordova-plugin-device.git`**

Reviewer 2: other variants

### Restoring Plugins

Plugins are automatically restored from `package.json` and `config.xml` when executing the the **`cordova prepare`** command.

If a plugin is defined in both files, the information defined in `package.json` is used as the source of truth.

After `prepare`, if a plugin was restored from `config.xml`, the `package.json` file will updated to reflect the restored plugin and will have identical vales from `config.xml`.

If you add a plugin without specifying a `<version | folder | git_url>`, the version that will be installed is taken from `package.json` or `config.xml`.

**If discovered** in both files, `package.json` is given higher priority over `config.xml`.

Example:

Suppose your `config.xml` file contains the following entry:

```xml
<?xml version='1.0' encoding='utf-8'?>
    ...
    <plugin name="cordova-plugin-device" spec="2.0.1" />
    ...
</ xml>
```

If you run the command **`cordova plugin add cordova-plugin-device`** with no `<version | folder | git_url>` specified, the platform `cordova-plugin-device@2.0.1` will be retrieved and installed.

> **Example Order of Priorty for Restoring Plugins:**

Reviewer 2: ```suggestion
**Example Order of Priority for Restoring Plugins:**
```

---

# Platforms and Plugins Version Management

> From version 4.3.0 onwards, Cordova provides the ability to save and restore platforms and plugins.

Reviewer 2: ```suggestion
Cordova provides the ability to save and restore platforms and plugins.
```

The **restore** step happens automatically when a **`cordova prepare`** is issued, making use of information previously saved in the `package.json` and `config.xml` files.

One scenario where save/restore capabilities come in handy is in large teams that work on an app, with each team member focusing on a platform or plugin. This feature makes it easier to share the project and reduce the amount of redundant code that is checked in the repository.

## Platform Versioning

### Saving Platforms

To save a platform, issue the following command:

```bash
cordova platform add <platform[@<version>] | directory | git_url>
```

After running the above command, the **`package.json`** should contain something as seen below:

```json
"cordova": {
  "platforms": [
    "android"
  ]
},
"dependencies": {
  "cordova-android": "^8.0.0",
}
```

The `--nosave` flag prevents adding and deleting specified platforms to the `package.json` file. To prevent saving a platform, issue the following command:

```bash
cordova platform add <platform[@<version>] | directory | git_url> --nosave
```

Some Examples:

* **`cordova platform add android`** 

>   Retrieves the pinned version of `cordova-android` platform from npm, adds it to the project and updates the `package.json` file.

Reviewer 2: config.xml is not updated here any more at all with the platform?

Author: As far as how I interpreted this commit, https://github.com/apache/cordova-lib/commit/8371d7c1bca6b4ca6e64055054405edf8ccf11b6

`config.xml` should no longer be updated when a plugin or platform is added or updated. `config.xml` will only be modified to remove the plugin or platform when the remove command is executed.

Reviewer 2: check @dpogue

Reviewer: That is correct, Cordova 9 does not add any platforms or plugins to config.xml. It only updates package.json.

* **`cordova platform add android@7.1.4`**

  Retrieves the `cordova-android` platform version `7.1.4` from npm, adds it to the project and updates the `package.json` file.

* **`cordova platform add https://github.com/apache/cordova-android.git`**

>   

Reviewer 2: why the empty lines here?

Author: In my previewer, without the extra empty lines, it would consolidate the lines together. I am not sure if this is how Visual Studio Code markdown previewer works or all previewer.

Reviewer 2: Ah, yeah you would have to add `  ` (to spaces) at the previous line so that it adds a `<br>` after the line. Markdown is strange that way.

  **`cordova platform add https://github.com/apache/cordova-android`**
  
  **`cordova platform add github:apache/cordova-android`**

  npm retrieves the `cordova-android` platform from the git repository, adds it to the project and updates the `package.json`.
  
* **`cordova platform add C:/path/to/android/platform`**

  Retrieves the Android platform from the specified directory, adds it to the project, and updates the `package.json` file.

* **`cordova platform add android --nosave`**

  Retrieves the pinned version of `cordova-android` platform from npm, adds it to the project, but does not add it to the `package.json` file.

### Updating or Removing Platforms

> It is also possible to update or delete from `config.xml` and `package.json` during the commands 'cordova platform update' and 'cordova platform remove' :

Reviewer 2: ```suggestion
It is also possible to update or delete from `config.xml` and `package.json` during the commands `cordova platform update` and `cordova platform remove`:
```

```bash
cordova platform update <platform[@<version>] | directory | git_url> --save
cordova platform remove <platform>
```

Some Examples:

* **`cordova platform update android --save`**

>   In addition to updating the `cordova-android` platform to the pinned version, it updates the `package.json` file.

Reviewer 2: Is "pinned" explained anywhere?

Author: No, as far as I know within docs, we do not explain what is and are the pinned platform versions for each CLI version.

* **`cordova platform update android@3.8.0 --save`**

  In addition to updating the `cordova-android` platform to version `3.8.0` it updates the `package.json` file.

* **`cordova platform update /path/to/android/platform --save`**

  In addition to updating the `cordova-android` platform to version in the folder it updates the `package.json` file.

* **`cordova platform remove android`**

>   Removes the `cordova-android` platform from the project and removes from the `package.json` file.

Reviewer 2: ```suggestion
  Removes the `cordova-android` platform from the project and removes it from the `package.json` file.
```

>   _Note: If the platform definiton existed in `config.xml` from a previous version of Cordova-CLI, it will also be removed from `config.xml`._

Reviewer 2: ```suggestion
  _Note: If the platform definiton existed in `config.xml` from a previous version of Cordova CLI, it will also be removed from `config.xml`._
```

* **`cordova platform remove android --nosave`**

  Removes the `cordova-android` platform from the project, but does not remove it from the `package.json` file.

### Restoring Platforms

> Platforms are automatically restored from the `package.json` and `config.xml` when executing the the **`cordova prepare`** command.

Reviewer 2: ```suggestion
Platforms are automatically restored from the `package.json` (and `config.xml`) when executing the the **`cordova prepare`** command.
```
Maybe to indicate that this is not the main mechanism any more?

> If a platfrom is defined in both files, the information defined in `package.json` is used as the source of truth.

Reviewer 2: ```suggestion
If a platform is defined in both files, the information defined in `package.json` is used as the source of truth.
```

> After `prepare`, if a platform was restored from `config.xml`, the `package.json` file will updated to reflect the restored platform and will have identical vales from `config.xml`.

Reviewer 2: ```suggestion
After `prepare`, if a platform was restored from `config.xml`, the `package.json` file will be updated to reflect the restored platform and will have identical values from `config.xml`.
```

> After `prepare`, if a platform was restored from `config.xml`, the `package.json` file will updated to reflect the restored platform and will have identical vales from `config.xml`.

Reviewer 2: ```suggestion
After `prepare`, if a platform was restored from `config.xml`, the `package.json` file will updated to reflect the restored platform and will have identical vales to `config.xml`.
```

If you add a platform without specifying a `<version | folder | git_url>`, the version that will be installed is taken from `package.json` or `config.xml`.

**If discovered** in both files, `package.json` is given higher priority over `config.xml`.

Example:

Suppose your `config.xml` file contains the following entry:

```xml
<?xml version='1.0' encoding='utf-8'?>
    ...
    <engine name="android" spec="7.1.4" />
    ...
</xml>
```

If you run the command **`cordova platform add android`** with no `<version | folder | git_url>` specified, the platform `android@7.1.4` will be retrieved and installed.

**Example Order of Priority for Restoring Platforms:**

Suppose you have defined in `config.xml` and `package.json` a platform and version as follows:

**`config.xml`**:

```xml
<engine name="android" spec=“7.4.1” />
```

**`package.json`**:

```json
"cordova": {
  "platforms": [
    "android"
  ]
},
"dependencies": {
  "cordova-android": "^8.0.0"
}
```

> When `prepare` is executed, the version from `package.json` has higher priority over  `config.xml` and version `^8.0.0` will be installed.

Reviewer 2: ```suggestion
When `prepare` is executed, the version from `package.json` has higher priority over `config.xml` and version `^8.0.0` will be installed.
```

---

## Plugin Versioning

> The plugin commands are a mirror of the platform commands

Reviewer 2: ```suggestion
The plugin commands are a mirror of the platform commands:
```

```bash
cordova plugin add <plugin[@<version>] | directory | git_url> --nosave
```

Some Examples:

* **`cordova plugin add cordova-plugin-device`**

  Retrieves the pinned version of the `cordova-plugin-device` plugin from npm, adds it to the project and updates the `package.json` file.

* **`cordova plugin add cordova-plugin-device@2.0.1`**

  Retrieves the `cordova-plugin-device` plugin at version `2.0.1` from npm, adds it to the project and updates the `package.json` file.

* **`cordova plugin add https://github.com/apache/cordova-plugin-device.git`**
  
  **`cordova plugin add https://github.com/apache/cordova-plugin-device`**
  
  **`cordova plugin add github:apache/cordova-plugin-device`**

  npm retrieves the `cordova-plugin-device` plugin from the git repository, adds it to the project and updates the `package.json`.

* **`cordova plugin add C:/path/to/console/plugin`**

  Retrieves the `cordova-plugin-device` plugin from the specified directory, adds it to the project, and updates the `package.json` file.

### Mass Saving of Plugins on an Existing Project

> If you have a pre-existing project and you want to save all currently added plugins in the project, you can use:

Reviewer 2: This is for when you used `nosave` before? Or how could those not be saved and require this command?

Author: This command was used when auto-saving was not a thing.

Since plugins and platforms are save automatically now, this command, in most cases, is not needed. 

As a possible current use case, if a user did use `--nosave`, to add the plugin(s), running the save command will add all the missing plugins to `config.xml`.

Anyways, this feature is to be removed. 
- This feature was already removed from platforms.
- The `--nosave` flag has issues. It removes dependencies and causes build failures do to missing dependencies. Using this flag is not normally recommended I beleive.
- This feature only updates `config.xml` which does not follow Cordova 9 plugin/platform management strategy. They should be stored in `package.json`

*`plugin save` Deprecate*
*Ticket:* https://github.com/apache/cordova-lib/issues/778
*PR:* https://github.com/apache/cordova-lib/pull/780

*`plugin save` Removal*
*Ticket:* https://github.com/apache/cordova-lib/issues/779
*PR:* https://github.com/apache/cordova-lib/pull/781

```bash
cordova plugin save
```

### Removing Plugins

> It is also possible to delete from `config.xml` and `package.json` during the command `cordova plugin remove`:

Reviewer 2: "also possible"? What does this refer to?

Author: The person who wrote this originally was stream lining from the previous section. The previous section talks about adding plugins so therefor it is also possible to remove.

I went ahead removed `also`, among other rewording changes.

```bash
cordova plugin remove <plugin>
```

For Example:

* **`cordova plugin remove cordova-plugin-device`**

  Removes the `cordova-plugin-device` plugin from the project and deletes its entry from `package.json`.

>   _Note: If the plugin definiton existed in `config.xml` from a previous version of Cordova-CLI, it will also be removed from `config.xml`._

Reviewer 2: ```suggestion
  _Note: If the plugin definiton existed in `config.xml` from a previous version of Cordova CLI, it will also be removed from `config.xml`._
```

If you add a plugin without specifying a `<version | folder | git_url>`, the version that will be installed is taken from `package.json` or `config.xml`.

**If discovered** in both files, `package.json` is given higher priority over `config.xml`.

Example:

Suppose your `config.xml` file contains the following entry:

```xml
<?xml version='1.0' encoding='utf-8'?>
    ...
    <plugin name="cordova-plugin-device" spec="2.0.1" />
    ...
</ xml>
```

If you run the command **`cordova plugin add cordova-plugin-device`** with no `<version | folder | git_url>` specified, the platform `cordova-plugin-device@2.0.1` will be retrieved and installed.

**Example Order of Priority for Restoring Plugins:**

Suppose you have defined in `config.xml` and `package.json` a plugin and version as follows:

**`config.xml`**:

```xml
<plugin name="cordova-plugin-splashscreen"/>
```

**`package.json`**:

```json
"cordova": {
  "plugins": [
    "cordova-plugin-splashscreen"
  ]
},
"devDependencies": {
  "cordova-plugin-splashscreen": "1.0.0"
}
```

> When `prepare` is executed, the version from `package.json` has higher priority over  `config.xml` and version `1.0.0` will be installed.

Reviewer 2: ```suggestion
When `prepare` is executed, the version from `package.json` has higher priority over `config.xml` and version `1.0.0` will be installed.
```

> * **`cordova platform update /path/to/android/platform --save`**

Reviewer 2: --save necessary here?

Author: I would say I can remove this. 

I don't remember adding `--save` so it might have been there from before or copy paste. 

Anyways, `cordova platform --help` also says nothing about `--save`, so I will remove.
